### DISCUSSION TITLE
Langium + ViteJS

### DISCUSSION THREAD

**Discussion by luiz290788** [eclipse-langium/langium#1009] [Q&A] [answered]
Is anyone using langium with vitejs? I'm trying to use and it works fine with the development environment, but when I try to build I am getting this error:

```
Uncaught TypeError: Object.defineProperty called on non-object
    at Function.defineProperty (<anonymous>)
```

Any clue on how to solve this?

**Comment by msujew** [comment]
@kaisalmen Any clue on this? If I remember correctly, you also had trouble using Langium with ViteJS in the context of the [langium-website](https://github.com/langium/langium-website) repo.

**Comment by kaisalmen** [comment, marked as answer]
Yes, when using vite build (=rollup) in lib mode for worker creation. We switched to esbuild with minfy.

**Reply by luiz290788** [reply]
I have the problem while building as well. I was able to build the worker, but if I import langium regardless of being the worker or in the main thread, I have that error. Any clue why that is happening? It happens in this part of the code:

```js
    function yb() {
        if (mb)
            return lu;
        mb = 1,
        Object.defineProperty(lu, "__esModule", {
            value: !0
        }),
        lu.createCompletionParser = void 0;
        const e = Cm()
          , t = hb();
        function r(n) {
            const i = n.Grammar
              , o = n.parser.Lexer
              , l = new e.LangiumCompletionParser(n);
            return (0,
            t.createParser)(i, l, o.definition),
            l.finalize(),
            l
        }
        return lu.createCompletionParser = r,
        lu
    }
```

The error happens at this line `Object.defineProperty(lu, "__esModule", {`. I checked where the `createCompletionParser` is declared and it doesn't seem to be doing anything too crazy there.

**Reply by kaisalmen** [reply]
It was working with vite 3 and stopped with vite 4 / rollup 3. We haven't found a solution and just changed the way the worker is created, for example: https://github.com/langium/langium-website/blob/main/hugo/package.json#L13
You can add `--minify` in that command to get a smaller output file

**Reply by luiz290788** [reply]
Was it the same problem that I'm facing or was it a different one?

**Reply by msujew** [reply]
I think it was the same issue, I remember reading about `createCompletionParser` in the error logs as well. It would be interesting to figure out what's going wrong there in rollup/vite, since all other bundlers don't seem to have this problem.

**Reply by luiz290788** [reply]
I'm trying to build directly with Rollup. I get this warning while building:

```
(!) Circular dependencies
../../node_modules/langium/lib/default-module.js -> ../../node_modules/langium/lib/grammar/grammar-config.js -> ../../node_modules/langium/lib/utils/grammar-util.js -> ../../node_modules/langium/lib/default-module.js
../../node_modules/langium/lib/generator/generator-node.js -> ../../node_modules/langium/lib/generator/node-processor.js -> ../../node_modules/langium/lib/generator/generator-node.js
../../node_modules/langium/lib/utils/grammar-util.js -> ../../node_modules/langium/lib/grammar/ast-reflection-interpreter.js -> ../../node_modules/langium/lib/grammar/type-system/ast-collector.js -> ../../node_modules/langium/lib/grammar/type-system/types-util.js -> ../../node_modules/langium/lib/grammar/type-system/type-collector/types.js -> ../../node_modules/langium/lib/generator/generator-node.js -> ../../node_modules/langium/lib/generator/node-processor.js -> ../../node_modules/langium/lib/generator/generator-tracing.js -> ../../node_modules/langium/lib/utils/grammar-util.js
...and 31 more
```

Could these circular dependencies be causing a module to be `undefined` before its use? Do we have these circular dependencies inside of Langium? Is it easy to remove those dependencies?

**Reply by luiz290788** [reply]
I checked the first one, it looks like there are circular dependencies indeed. I'm not sure if they are what is causing the problem.

**Reply by msujew** [reply]
Circular dependencies in JS usually aren't a problem unless they are load time evaluated. Something like:

```js
import { x } from './x';
export const y = 5;
console.log(x);
```

```js
import { y } from './y';
export const x = 5;
console.log(y);
```

These type of imports will always lead to issues (both bundled and plain commonJS), but everything else should be fine.

**Reply by luiz290788** [reply]
Sorry, not sure if I follow, you are saying that it shouldn't be the problem in my case, or it would be the problem because I'm trying to bundle?

**Reply by msujew** [reply]
It shouldn't be a problem, since we only use circular dependencies which bundlers and runtimes should be easily able to deal with. Though we should look into whether some of these are truly necessary. Most of them are probably just type imports which aren't declared as such.

**Reply by luiz290788** [reply]
I don't believe that the problem is the circular dependencies. I couldn't isolate the problem, I don't if it is something that was done in Langium or if it is a bug in Vite/Rollup. I will probably use Vite for the development and try to build with esbuild just for me to able to deploy my application, and then I will revisit this in the future.

**Reply by msujew** [reply]
I don't think so either. I also don't think it's something that Langium is doing (as at least nothing that's fundamentally wrong), since other bundlers, like esbuild and webpack, and the node runtime itself can deal just fine with it. It's likely just something that Langium is doing that Rollup somehow isn't able to deal with, though I can't think of anything really.

**Reply by luiz290788** [reply]
Yeah, I agree. I opened a bug for Vite. It could also be related to the specific configuration they apply to rollup.

**Reply by luiz290788** [reply]
One interesting finding, if try to bundle the langium package in the langium codebase using rollup it works fine. I don't know if the problem is related to the transpiled code.

**Reply by luiz290788** [reply]
There is a [solution](https://github.com/vitejs/vite/issues/12660#issuecomment-1490578893), it is to set `strictRequires` to `false` in the `commonjsOptions`.  Look at the documentation of [strictRequires](https://github.com/rollup/plugins/tree/master/packages/commonjs/#strictrequires), it looks like that the problem has to do with the circular dependencies inside of langium. I'm happy to set this to false and continue with my work, as a nice to have, if you could remove the dependencies in future releases, that would help.

**Reply by msujew** [reply]
Ha, thanks for investigating further, that is really good to know. We'll take some time in the future to remove those cyclic imports, to make it easier to consume Langium in Vite 👍
